cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -225,19 +270,21 @@ protected void run() {
     // Clear the interrupt flag
     Thread.interrupted();
 
-    if (stopped) {
+    if (killed) {
       // TODO: Kill the remote program execute on explicit stop on this monitor
     }
 
-    // Publish the remaining program state messages. We should retry indefinitely
-    String topicConfig = Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC;
-    String topic = requestKeyToLocalTopic.get(topicConfig);
-    try {
-      Retries.callWithRetries(() -> publish(topicConfig, topic, lastProgramStateMessages),
-                              RetryStrategies.fixDelay(1, TimeUnit.SECONDS), Retries.ALWAYS_TRUE);
-    } catch (Exception e) {
-      // This shouldn't happen since we have infinite retry.
-      LOG.error("Failed to publish remaining program state messages for program run {}", programRunId, e);
+    if (lastProgramStateMessages.size() > 0) {
+      // Publish the remaining program state messages. We should retry indefinitely
+      String topicConfig = Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC;
+      String topic = requestKeyToLocalTopic.get(topicConfig);
+
+      callWithRetries((Retries.Callable<Long, TransactionFailureException>) () -> {
+        Transactionals.execute(transactional, context -> {
+          return publish(topicConfig, topic, lastProgramStateMessages, context);
```

Comment:
In hindsight, probably not a good idea to retry indefinitely as it can block the stopAndWait call.